### PR TITLE
Some more service page optimizations

### PR DESCRIPTION
**Pull request by KKoukiou**

### CODE REVIEW

File: pkg/systemd/services/services-list.jsx [hunk 1]
```diff
@@ -40,38 +40,39 @@ export const ServicesList = ({ units, isTimer }) => {
...
 };
 
-const ServicesRow = ({ Id, shortId, AutomaticStartup, UnitFileState, LoadState, HasFailed, CombinedState, LastTriggerTime, NextRunTime, Description, WantedBy, isTimer }) => {
-    const props = { shortId, Description };
-    const columnsMap = {
-        shortId: { value: _("Name"), className: "service-unit-id", width: 2 },
-        Description: { value: _("Description"), className: "service-unit-description", width: 4 },
-        Triggers: { value: _("Triggers"), timerOnly: true, className: "service-unit-triggers", width: 3 },
-    };
+class ServicesRow extends React.PureComponent {
```
Comment (martinpitt) [inline pkg/systemd/services/services-list.jsx:43]: Feels like a step backwards to have to convert from a simple function to a class -- but apparently [pure functional components](https://blog.logrocket.com/pure-functional-components-in-react-16-6/) need another npm module..

Comment (KKoukiou) [reply pkg/systemd/services/services-list.jsx:43]: I can drop that commit if you don't prefer it - I just wanted to try it out basically.
I don't want to bring in a new NPM module just for this.

Comment (martinpitt) [reply pkg/systemd/services/services-list.jsx:43]: Sorry, that wasn't meant as a veto, more like a "sigh" :-)

**Comment by martinpitt** [comment]
The remaining failures look like scattered flakes, retrying for good measure. The stress test apparently succeeded everywhere, amazing work! :hugs: 

I'll review this in more detail on Monday.

**Comment by martinpitt** [comment]
[Services page crashes on Edge](https://logs.cockpit-project.org/logs/pull-13958-20200425-103800-f8c4c911-fedora-31-selenium-edge/log.html#4), but it already does that on master, so let's fix that in a follow-up.

**Comment by martinpitt** [comment]
services test failures on [fedora-32](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-13958-20200425-103059-f8c4c911-fedora-32/log.html#252-2) and [rhel-8-2](https://logs.cockpit-project.org/logs/pull-13958-20200425-103057-f8c4c911-rhel-8-2/log.html#251-2), but I think at least the second one is unrelated - it checks pkg/lib/service.js, not the services page.

File: pkg/systemd/services/services.jsx [hunk 2]
```diff
@@ -422,7 +425,19 @@ class ServicesPage extends React.Component {
...
+                                            for (const unitPath in this.state.unit_by_path) {
+                                                if (!this.seenPaths.has(unitPath)) {
+                                                    hasExtraEntries = true;
+                                                    delete unit_by_path[unitPath];
+                                                }
+                                            }
+                                            if (hasExtraEntries)
+                                                this.setState({ unit_by_path });
+
+                                            this.setState({ loadingUnits: false, isFullyLoaded: true });
```
Comment (martinpitt) [inline pkg/systemd/services/services.jsx:438]: Could these two setState calls be combined into one? Like, assign the new state to a variable, and extend it `if hasExtraEntries`.

File: pkg/systemd/services/services.jsx [hunk 3]
```diff
@@ -567,7 +572,7 @@ class ServicesPage extends React.Component {
             unit.shortId = unit.Id.substring(0, unit.Id.length - 8);
     }
 
-    updateProperties(props, path) {
+    updateProperties(props, path, updateFileState = false) {
```
Comment (martinpitt) [inline pkg/systemd/services/services.jsx:568]: Please no spaces around `=` for default arguments

Comment (KKoukiou) [reply pkg/systemd/services/services.jsx:568]: Oh this is forced by eslint:
/home/kkoukiou/repos/cockpit/pkg/systemd/services/services.jsx
  568:50  error  Operator '=' must be spaced  space-infix-ops

Comment (KKoukiou) [reply pkg/systemd/services/services.jsx:568]: Nope - they don't support this distinction  - sorry :( 
https://github.com/eslint/eslint/issues/7489

Comment (martinpitt) [reply pkg/systemd/services/services.jsx:568]: Nevermind then :)

File: pkg/systemd/services/services-list.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/systemd/services/services-list.jsx:43]: I don't think we should have the rule to use functions if at all possible. Always using a class is more uniform and we don't have to flip back and forth between using a function and using a class depending on whether a component has state or not, or derives from PureComponent or not, or needs some lifecycle methods or not.

File: pkg/systemd/services/services.jsx [hunk 4]
```diff
@@ -405,9 +418,8 @@ class ServicesPage extends React.Component {
                                         .then(result => {
                                             // First add the path in the seens paths to maintain the loading state
                                             result.forEach(path => this.seenPaths.add(path));
-                                            const allPaths = paths.concat(result);
 
-                                            return Promise.all(allPaths.map(path => this.getUnitByPath(path)));
+                                            return Promise.all(result.map(path => this.getUnitByPath(path)));
```
Comment (mvollmer) [inline pkg/systemd/services/services.jsx]: Why are there calls to LoadUnit and getUnitByPath at all in listUnits? Wasn't the goal to get rid of all of them?

Comment (KKoukiou) [reply pkg/systemd/services/services.jsx]: For the units that are not listed by ListUnits but only by ListUnitFiles, we still need to get some information for these - like Description to show in the main page.

**Comment by KKoukiou** [comment]
Hmm https://logs.cockpit-project.org/logs/pull-13958-20200427-162819-1e9c6754-fedora-32/log.html failed to times in a row - seems like I introduced a flake somehow :confused:

**Pull request merged**